Kernel/Net/NetworkAdapter.cpp

```diff
@@ -76,15 +76,15 @@ void NetworkAdapter::send(const MACAddress& destination, const ARPPacket& packet
     send_raw({ (const u8*)eth, size_in_bytes });
 }
 
-KResult NetworkAdapter::send_ipv4(const IPv4Address& source_ipv4, const MACAddress& destination_mac, const IPv4Address& destination_ipv4, IPv4Protocol protocol, const UserOrKernelBuffer& payload, size_t payload_size, u8 ttl)
+void NetworkAdapter::fill_in_ipv4_header(PacketWithTimestamp& packet, IPv4Address const& source_ipv4, MACAddress const& destination_mac, IPv4Address const& destination_ipv4, IPv4Protocol protocol, size_t payload_size, u8 ttl)
 {
     size_t ipv4_packet_size = sizeof(IPv4Packet) + payload_size;
-    if (ipv4_packet_size > mtu())
-        return send_ipv4_fragmented(source_ipv4, destination_mac, destination_ipv4, protocol, payload, payload_size, ttl);
+    VERIFY(ipv4_packet_size <= mtu());
 
     size_t ethernet_frame_size = sizeof(EthernetFrameHeader) + sizeof(IPv4Packet) + payload_size;
-    auto buffer = NetworkByteBuffer::create_zeroed(ethernet_frame_size);
-    auto& eth = *(EthernetFrameHeader*)buffer.data();
+    VERIFY(packet.buffer.size() == ethernet_frame_size);
+    memset(packet.buffer.data(), 0, sizeof(EthernetFrameHeader) + sizeof(IPv4Packet));
```

Comment:
Use your fancy `layer3_payload_offset()`? 
```suggestion
    memset(packet.buffer.data(), 0, layer3_payload_offset());
```

Reply:
Fixed and fixed. :)